### PR TITLE
ENH: enables `merge-taxa` and `dereplicate` to handle user-defined rank handles.

### LINKED ISSUE #122: IMP: update dereplicate backfilling functionality
Update the `dereplicate` action to handle any number or combination of taxonomic ranks the user would like to use when implementing any of the LCA approaches. For the moment, we can only handle dealing with a standard 7-rank taxonomy, *i.e.* dpcofgs.

An example of this issue is on the [QIIME 2 Forum](https://forum.qiime2.org/t/qiime-rescript-dereplicate-p-mode-super-leads-to-taxonomic-label-depth-is-uneven-error-with-qiime-rescript-evaluate-fit-classifier/21791/2).

### PR DESCRIPTION
**Pull request by mikerobeson**
This PR should allow `merge-taxa` and `dereplicate` to handle user-defined rank handles. This is an attempt to fix #122.

- `merge-taxa`:
  - `--p-new-rank-handles`: is now a list of allowed ranks (*e.g.* `parse-silva-taxonomy`, `get-ncbi-data`)
  - ~~`--p-apply-new-rank-handles`: is a boolean to apply `--p-new-rank-handles`~~
- `dereplicate`:
  - `--p-rank-handles`: is now a list of allowed ranks (*e.g.* `parse-silva-taxonomy`, `get-ncbi-data`)
  - ~~`--p-backfill`: is a boolean to apply `--p-rank-handles`~~
  
 To do:
  - [x] Use taxonomy_rank to taxonomy_prefix dictionary, *i.e.* {'domain' : 'd__', ...},  to maintain constancy with `parse-silva-taxonomy`, `get-ncbi-data`

### CODE REVIEW

**Review by nbokulich — CHANGES_REQUESTED** [review]
thanks @mikerobeson looking good! I made a few comments throughout... pls see what you think!

File: rescript/dereplicate.py [hunk 1]
```diff
@@ -13,16 +13,20 @@
...
 
 def dereplicate(sequences: DNAFASTAFormat,
                 taxa: pd.DataFrame,
                 mode: str = 'uniq',
                 perc_identity: float = 1.0,
                 threads: int = 1,
-                rank_handles: str = 'silva',
+                rank_handles: list = ['domain', 'phylum', 'class', 'order',
+                                      'family', 'genus', 'species'],
+                backfill: bool = True,
```
[the remark this answers is not in the document]
Comment (mikerobeson) [reply rescript/dereplicate.py:29]: I think I had originally thought of that, but was unsure of out how to apply that approach in this context. Given that we expect a list. Do I just simply search for the string `disable` in the list? Thus, even if there are other items provided (say by accident) it'll ignore them. Or did you have something else in mind?

File: rescript/dereplicate.py [hunk 2]
```diff
@@ -13,16 +13,20 @@
...
+from .ncbi import _allowed_ranks
+
 
 def dereplicate(sequences: DNAFASTAFormat,
                 taxa: pd.DataFrame,
                 mode: str = 'uniq',
                 perc_identity: float = 1.0,
                 threads: int = 1,
-                rank_handles: str = 'silva',
+                rank_handles: list = ['domain', 'phylum', 'class', 'order',
```
[the remark this answers is not in the document]
Comment (mikerobeson) [reply rescript/dereplicate.py:27]: I think I am actually using `_allowed_ranks` from `ncbi.py`. I should have just called values() on that instead of using `_allowed_rank_handles`. Though I did notice a mistake in [_allowed_ranks](https://github.com/mikerobeson/RESCRIPt/blob/bdd71bd0f2ebc8ba25055d59315b7f623112926e/rescript/ncbi.py#L39), both `family` and `forma` use `f__` as the rank handle.

File: rescript/dereplicate.py [hunk 3]
```diff
@@ -53,21 +57,25 @@ def dereplicate(sequences: DNAFASTAFormat,
...
 
 def _backfill_taxonomy(taxon, rank_handles):
-    formatted_taxon = taxon.split(';')
+    formatted_taxon = [t.strip() for t in taxon.split(';')]
     tax_len = len(formatted_taxon)
     if tax_len >= len(rank_handles):
         return taxon
     else:
-        return ';'.join(formatted_taxon + rank_handles[tax_len:])
+        return '; '.join(formatted_taxon + rank_handles[tax_len:])
```
[the remark this answers is not in the document]
Comment (mikerobeson) [reply rescript/dereplicate.py:78]: Yeah, I was going back and forth on this myself. I can retain the `;` w/o the space.

File: rescript/merge.py [hunk 4]
```diff
@@ -72,10 +73,14 @@ def merge_taxa(data: pd.DataFrame,
             result = result[reordered_cols]
 
     # Insert new rank handles if selected
-    if new_rank_handle is not None:
-        new_rank_handle = _rank_handles[new_rank_handle]
+    if apply_new_rank_handles:
+        # Sort user ranks relative to allowed_ranks, this ensures
+        # that the taxonomic ranks supplied by the user are in order
+        sorted_rank_handles = [p for r, p in _allowed_ranks.items()
```
[the remark this answers is not in the document]
Comment (mikerobeson) [reply rescript/merge.py:79]: Right! Will do.

File: rescript/plugin_setup.py [hunk 5]
```diff
@@ -241,17 +240,18 @@
...
             'only contain as many levels as there are handles if this '
-            'parameter is used. So "greengenes" will trim all taxonomies to '
-            'seven levels, even if longer annotations exist. Note that this '
+            'parameter is used. This will trim all taxonomies to the given '
+            'levels, even if longer annotations exist. Note that this '
             'parameter will prepend rank handles whether or not they already '
             'exist in the taxonomy, so should ALWAYS be used in conjunction '
             'with `rank_handle_regex` if rank handles exist in any of the '
             'inputs.'),
+        'apply_new_rank_handles': 'Apply \'new-rank-handles\' to taxonomy '
```
[the remark this answers is not in the document]
Comment (mikerobeson) [reply rescript/plugin_setup.py:253]: Ohh, good catch, I missed that!

File: rescript/dereplicate.py [hunk 6]
```diff
@@ -53,21 +57,25 @@ def dereplicate(sequences: DNAFASTAFormat,
             derep_taxa, seqs_out = _dereplicate_taxa(
                 taxa, sequences, clustered_seqs, uc, mode=mode)
 
-            if rank_handles != 'disable':
-                rank_handles = _rank_handles[rank_handles]
+            if backfill:
+                # Sort user ranks relative to allowed_ranks, this ensures
```
[the remark this answers is not in the document]
Comment (nbokulich) [reply rescript/dereplicate.py:61]: yeah that's what I am wondering... but I like the way you have this for grabbing the handles from a dict and reinforcing order (just move to a separate function to reduce code duplication in merge and dereplicate)

File: rescript/dereplicate.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nbokulich) [reply rescript/dereplicate.py:27]: okay understood!

File: rescript/dereplicate.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mikerobeson) [reply rescript/dereplicate.py]: I left a note / question in the `test_dereplicate.py` file. I think we should add a user defined delimiter, like `;` or `; `. In case the user would like to keep the spacing between the ranks consistent when backfilling.

File: rescript/dereplicate.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mikerobeson) [reply rescript/dereplicate.py:29]: Ohh, great idea! I'll give that a try.

File: rescript/dereplicate.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mikerobeson) [reply rescript/dereplicate.py]: >maybe set up a function to autodetect how taxonomy strings should be formatted?

I was just thinking that I can simply search for variations of the `;` delimiter with and without spaces, then use the most abundant one for backfilling.

> (on the other hand we could just have all rescript actions always remove any leading/trailing spaces in all taxonomy strings,

In the prior iteration of this code, I was stripping space off of each taxonomic rank, and re-joining. But I see your point about cryptic issues. And the work of updating all the other rescript actions...

Or am I making much ado about nothing here? I'll think on it...

[the remark this answers is not in the document]
Comment (nbokulich) [reply rescript/dereplicate.py]: this might be a problem that we just can't fix (as issues like joining with other databases, etc, are something that users should exercise due diligence over, and not something that we can control).

I am happy to just stick with the status quo to strip from all taxonomic ranks, and we could open an issue to provide a uniform solution across RESCRIPt if you think that we should provide the option. 

Just to say, I think this is why in the original implementation of `merge` I did not expose the ranks themselves but the "flavor" of taxonomy that the user wanted to use. Then the formatting comes pre-baked. Obviously only a half-baked solution 😂 , hence #122 .

[the remark this answers is not in the document]
Comment (mikerobeson) [reply rescript/dereplicate.py]: Or better yet, perhaps we can simply point users to `edit-taxonomy` or wrap that action into this action prior to it returning output?

[the remark this answers is not in the document]
Comment (nbokulich) [reply rescript/dereplicate.py]: yeah I like that approach (documenting that spaces are stripped, and pointing to `edit-taxonomy`). I would not wrap the action itself (as that would entail always passing over the entire series an additional time), but the internal function could be called to edit strings inline if internal spaces are detected, if you like the autodetect idea.

[the remark this answers is not in the document]
Comment (mikerobeson) [reply rescript/dereplicate.py]: @nbokulich For the moment, I decided to not worry about the inconstancy of spaces surrounding `;`. We can discuss further offline if interested.

File: rescript/merge.py [hunk 7]
```diff
@@ -22,7 +22,8 @@
 def merge_taxa(data: pd.DataFrame,
                mode: str = 'len',
                rank_handle_regex: str = '^[dkpcofgs]__',
-               new_rank_handle: str = None,
```
Comment (nbokulich) [inline rescript/merge.py:25]: did you intend to change the default? I am fine with this, but just want to make sure — it should be documented in the release notes in the next tagged release of rescript

File: rescript/plugin_setup.py [hunk 8]
```diff
@@ -242,17 +241,17 @@
...
             'only contain as many levels as there are handles if this '
-            'parameter is used. So "greengenes" will trim all taxonomies to '
-            'seven levels, even if longer annotations exist. Note that this '
+            'parameter is used. This will trim all taxonomies to the given '
+            'levels, even if longer annotations exist. Note that this '
             'parameter will prepend rank handles whether or not they already '
             'exist in the taxonomy, so should ALWAYS be used in conjunction '
             'with `rank_handle_regex` if rank handles exist in any of the '
-            'inputs.'),
+            'inputs. Use \'disable\' to prevent applying \'new_rank_handles\''
```
Comment (nbokulich) [inline rescript/plugin_setup.py:253]: applying --> appending

(just to make terminology consistent with the previous sentence)

**Review by nbokulich — CHANGES_REQUESTED** [review]
Thanks @mikerobeson ! I like it, and tested with some greengenes data.

One more change is needed, and also one more comment:
1. Could you add some tests for non-7-level taxonomies? Since I think that was the motivation for this PR but the current tests are all 7-level so use the defaults. The tests I can think of: 1-level, 3-level, 9-level, all-levels (just to test a few normal and edge cases),  one test where the new rank handles are a subset of the original, and one case where the new rank handles include a rank not found in the original. The labels presumably do not matter, so this could be a toy example, no need to assemble real data, though one of the motivating examples from non-bacterial taxonomies could be a good test.
2. comment: the inconsistent spacing is still making me worry just a little a bit. I wonder if all rescript actions (or at least dereplicate) should just strip spacing by default, or if these actions should autodetect and add back spaces. This could be a separate PR, but I feel like it is the sort of thing that users would complain about and should be fixed before we tag another release.

**Comment by mikerobeson** [comment]
I started to revisit this, at least the `dereplicate.py` code. I think I have an acceptable approach to this. I decided to normalize the taxonomy strings by allowing a user option (`--p-coerce-sc-delim`) for the semicolon. That is, they have a choice of `;`, `; `, or `  ;  `.   I may change this to `--p-semicolon-spacing`, with options for `space before`, `space after`, `both`. 🤷‍♂️ 

This take affect before any other processing is done. 

Anyway, I still need to add the other tests you requested, but I wanted to implement this first.

**Comment by nbokulich** [comment]
Hi @mikerobeson ,
I am not sure that we want to expose another parameter, and especially one that could lead to mistakes. Technically other delimiters could be passed (which could be useful in places, but could also cause other issues), and we create complication one way or another (what is a reasonable default? if none then users have another pesky param to pass every time they run this. if a default, then users may overlook, leading to other possible issues downstream).

I think I am fine with coercing a specific delimiter as you have done, but maybe we just always strip leading and tailing spaces — I think that's what we agreed on above?

**Comment by mikerobeson** [comment]
Alright. I removed the parameter option for the semicolons, and left a note in the help text to use `rescript edit-taxonomy` if the user would like to change the delimiter formatting. I also fixed / updated the test code. I'll work on adding the longer / shorter taxonomy string tests next.

File: rescript/tests/test_derep.py [hunk 9]
```diff
@@ -349,3 +436,20 @@ def _backfill_series(series, rank_handles=default_rank_handle):
...
+                            ' g__; s__; ssb__; for__')
+        f_taxa = self.fungi_taxa.view(pd.Series).sort_index()
+        f_taxa = f_taxa.apply(lambda x: ';'.join(
+                          _return_stripped_taxon_rank_list(x)))
+        backfilled_f_taxa = _backfill_series(f_taxa, alt_rank_handles)
+        trimmed_f_taxa = backfilled_f_taxa.apply(
+            lambda x: ';'.join(x.split(';')[:9]))
+        # manually backfill
+        exp_f_taxa = trimmed_f_taxa.apply(lambda x: x + ';g__;s__;ssb__;for__')
+        updated_backfilled_f_taxa = _backfill_series(trimmed_f_taxa,
```
Comment (nbokulich) [inline rescript/tests/test_derep.py:452]: hey @mikerobeson this new test is a bit confusing. I am not following the logic... you backfill, trim, then backfill again. And the expected taxonomy is builf off of the taxonomy that was backfilled once. Could you do the following to make this test a bit more transparent?
1. Move to its own unit test so that any failures are more transparently reported.
2. explicitly write out the expected taxonomy instead of building this off of backfilled data.
3. there are quite a few extra unneeded lines here because you test on a series. You could instead just test on a list of, say, 3 taxonomy strings of different lengths and contents (or however many edge cases you see) and this would eliminate the need for several lines of code.

**Review by nbokulich — CHANGES_REQUESTED** [review]
thanks @mikerobeson ! I like the new unit tests, except for the last one, which looks rather over-complicating for the backfill taxonomy function. Please see my comment in-line. Getting close!

File: rescript/tests/test_derep.py (hunk 9 above)
Comment (mikerobeson) [reply rescript/tests/test_derep.py]: HI @nbokulich, looking at this again with fresh eye... yeah... not sure what I was thinking. 🤦‍♂️ 

Hopefully, this new refactored version is in line with what you suggested.

**Review by nbokulich — APPROVED** [review]
now that's a nice clear test, thanks @mikerobeson !

I am very happy to say that all LGTM, thanks for bearing with me on this on putting together such an excellent enhancement! I'm looking forward to using this on some data now... 😁
